Warn on all missing fields in Cargo.toml

When using wasm-pack via `wasm-pack init` only one field at a time is output as to what's missing from the Cargo.toml file. It would be nicer if all the fields missing were printed out all at once!

I also found it difficult to know what it was telling me.

It would be good if it said something like "[ERROR] The following fields are missing from your `Cargo.toml`: repository, license"

I honestly thought that it was just giving me a warning, and spent 10 minutes ignoring the message, and wondering why it wasn't working... >_<

--- a/src/manifest.rs
+++ b/src/manifest.rs
@@ -4,6 +4,7 @@ use std::io::prelude::*;
 use console::style;
 use emoji;
 use failure::Error;
+use indicatif::MultiProgress;
 use progressbar;
 use serde_json;
 use toml;
@@ -17,20 +18,20 @@ struct CargoManifest {
 struct CargoPackage {
     name: String,
     authors: Vec<String>,
-    description: String,
+    description: Option<String>,
     version: String,
-    license: String,
-    repository: String,
+    license: Option<String>,
+    repository: Option<String>,
 }
 
 #[derive(Serialize)]
 struct NpmPackage {
     name: String,
     collaborators: Vec<String>,
-    description: String,
+    description: Option<String>,
     version: String,
-    license: String,
-    repository: Repository,
+    license: Option<String>,
+    repository: Option<Repository>,
     files: Vec<String>,
 }
 
@@ -64,10 +65,10 @@ impl CargoManifest {
             description: self.package.description,
             version: self.package.version,
             license: self.package.license,
-            repository: Repository {
+            repository: self.package.repository.map(|repo_url| Repository {
                 ty: "git".to_string(),
-                url: self.package.repository,
-            },
+                url: repo_url,
+            }),
             files: vec![js_file, wasm_file],
         }
     }
@@ -80,14 +81,41 @@ pub fn write_package_json(path: &str, scope: Option<String>) -> Result<(), Error
         style("[4/7]").bold().dim(),
         emoji::MEMO
     );
-    let pb = progressbar::new(step);
+
+    let warn = |field| {
+        format!(
+            "{} {}: Field {} is missing from Cargo.toml. It is not necessary, but recommended",
+            emoji::WARN,
+            style("[WARN]").bold().dim(),
+            field
+        )
+    };
+
+    let m = MultiProgress::new();
+    let pb = m.add(progressbar::new(step));
+
     let pkg_file_path = format!("{}/pkg/package.json", path);
     let mut pkg_file = File::create(pkg_file_path)?;
     let crate_data = read_cargo_toml(path)?;
     let npm_data = crate_data.into_npm(scope);
+
+    if npm_data.description.is_none() {
+        let warn_pb = m.add(progressbar::new(warn("description")));
+        warn_pb.finish();
+    }
+    if npm_data.repository.is_none() {
+        let warn_pb = m.add(progressbar::new(warn("repository")));
+        warn_pb.finish();
+    }
+    if npm_data.license.is_none() {
+        let warn_pb = m.add(progressbar::new(warn("license")));
+        warn_pb.finish();
+    }
+
     let npm_json = serde_json::to_string_pretty(&npm_data)?;
     pkg_file.write_all(npm_json.as_bytes())?;
     pb.finish();
+    m.join_and_clear()?;
     Ok(())
 }
 

--- a/tests/fixtures/bad-cargo-toml/Cargo.toml
+++ b/tests/fixtures/bad-cargo-toml/Cargo.toml
@@ -1,0 +1,7 @@
+[package]
+name = "bad-cargo-toml"
+version = "0.1.0"
+authors = ["Michael Gattozzi <mgattozzi@gmail.com>"]
+
+[dependencies]
+wasm-bindgen = "0.1"

--- a/tests/fixtures/bad-cargo-toml/README.md
+++ b/tests/fixtures/bad-cargo-toml/README.md
@@ -1,0 +1,2 @@
+# bad-cargo-toml
+> an example rust -> wasm project

--- a/tests/fixtures/bad-cargo-toml/src/lib.rs
+++ b/tests/fixtures/bad-cargo-toml/src/lib.rs
@@ -1,0 +1,15 @@
+#![feature(proc_macro)]
+
+extern crate wasm_bindgen;
+
+use wasm_bindgen::prelude::*;
+
+#[wasm_bindgen]
+extern {
+    fn alert(s: &str);
+}
+
+#[wasm_bindgen]
+pub fn greet(name: &str) {
+    alert(&format!("Hello, {}!", name));
+}

